### PR TITLE
Command line argument to configure the key establishment algorithm

### PR DESCRIPTION
**Pull request by colmenero**
The ability to configure the key establishment algorithm may be useful for additional testing between vendors.

This PR adds the command line parameter. It also adds the configuration for RTI (but not for other vendors).

### CODE REVIEW

**Review by ClarkTucker — APPROVED** [review]
Looks good to me!

File: srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h
Comment (YushengYangRTI) [inline srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h]: Check return value.

File: srcCxx/ShapeType_main.cxx
Comment (YushengYangRTI) [inline srcCxx/ShapeType_main.cxx]:
```suggestion
               governance_file, permissions_file,
               key_establishment_algorithm,
```

File: srcCxx/ShapeType_main.cxx
Comment (YushengYangRTI) [inline srcCxx/ShapeType_main.cxx]:
```suggestion
  printf( "    [-kest <authentication.key_establishment_algorithm>]    :  default: '%s'\n", DEFAULT_KEY_EST_ALGORITHM);
```
so that people can search for this property in [our documentation](https://community.rti.com/static/documentation/connext-dds/7.3.0/doc/manuals/connext_dds_secure/users_manual/p2_core/authentication.html#rti-sp-product-heading-properties-for-configuring-authentication).

**Review by YushengYangRTI — CHANGES_REQUESTED** [review]
https://github.com/omg-dds/dds-security/pull/16#discussion_r1668949455

File: srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h
Comment (colmenero) [reply srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h]: I added checks for the return value of all assertions

File: srcCxx/ShapeType_main.cxx
Comment (colmenero) [reply srcCxx/ShapeType_main.cxx]: @YushengYangRTI This code is common to all vendors, so I don't think we should point to a rti-specific property here.

Comment (ClarkTucker) [reply srcCxx/ShapeType_main.cxx]: We could use "dds.sec.auth.key_establishment_algorithm" instead of "authentication. key_establishment_algorithm".
The v1.2 spec has formalized that name.

File: srcCxx/ShapeType_main.cxx
Comment (YushengYangRTI) [inline srcCxx/ShapeType_main.cxx]:
```suggestion
                fprintf(stderr, "Error: missing <value of dds.sec.auth.key_establishment_algorithm> after \"-kest\"\n");
```

File: srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h
Comment (YushengYangRTI) [inline srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h]: I think we're leaking `pQos` in general.
```suggestion
            goto done;
```
and finalize `pQos` after `done:`.

File: srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h
Comment (YushengYangRTI) [inline srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h]:
```suggestion
        if (DDS_PropertyQosPolicyHelper_assert_property(
                &pQos.property,
```
Check all occurrences.

File: srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h
Comment (YushengYangRTI) [inline srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h]:
```suggestion
                DDS_BOOLEAN_FALSE) != DDS_RETCODE_OK) {
```

**Review by YushengYangRTI — CHANGES_REQUESTED** [review]
don't leak participantQos

File: srcCxx/ShapeType_main.cxx
Comment (mitza-oci) [inline srcCxx/ShapeType_main.cxx]: Since this changes the API between the common code in ShapeType_main.cxx and the implementation-specific code in the subdirectories, I suggest having this PR update the code for other (non-RTI) implementations so that their `create_participant` functions take the additional `const char*` parameter and ignore it.  Otherwise this is going to be a breaking change.  I realize you don't have development environments set up for all other implementations, but this should be a trivial fix.

File: srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h
Comment (colmenero) [reply srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h]: **Note to self**: This is  the C `DDS_DomainParticipantQos` structure and the traditional C++ APIs that are called. It is not the `DomainParticipantQos` modern C++ object which should be destructed when going out of scope.

File: srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h
Comment (YushengYangRTI) [inline srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h]: If `participant != NULL`, don't we still need to finalize the `pQos`?

File: srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h
Comment (YushengYangRTI) [inline srcCxx/rti_connext_dds/ShapeType_configurator_rti_connext_dds.h]: If `dp == NULL`, don't we still want to finalize the factory?

**Review by YushengYangRTI — CHANGES_REQUESTED** [review]

**Review by mitza-oci — APPROVED** [review]

**Review by YushengYangRTI — APPROVED** [review]

**Comment by colmenero** [comment]
@ClarkTucker If you agree, I'm holding off the merge until you resolve conflicts and merge the changes [in your PR](https://github.com/omg-dds/dds-security/pull/17) (which is based in the `feature/DH` branch) to this one. Then we can merge both to master here.

Or you can base your changes on master and merge the two PRs separately. That would be cleaner, but it will require some work. I'm leaving it up to you to decide.

Thanks!

**Comment by ClarkTucker** [comment]
I think the merge with the feature/automation branch is resolved (i've updated that branch).

I believe you should be able to merge this (feature/dh) branch into master.
Then I can merge the feature/automation branch into master; and that will show up as two different merges on master.

[I'm a git novice, tho, so I'm open to suggestions.]

**Comment by mitza-oci** [comment]
> I think the merge with the feature/automation branch is resolved (i've updated that branch).
> 
> I believe you should be able to merge this (feature/dh) branch into master. Then I can merge the feature/automation branch into master; and that will show up as two different merges on master.
> 
> [I'm a git novice, tho, so I'm open to suggestions.]

I updated the base branch over on #17 so it will be merged to master when someone hits that button

**Pull request merged**
